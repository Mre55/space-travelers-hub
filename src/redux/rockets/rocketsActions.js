import axios from 'axios';
// *Constants

// export const FETCH_ROCKETS = 'FETCH_ROCKETS';

// *Actions

// // import { FETCH_ROCKETS } from './rocketsActions';

// // Fetch data from API

export function fetchRocketSuccess(payload) {
  return {
    type: 'FETCH_ROCKETS',
    payload,
  };
}

export function fetchRockets() {
  return (dispatch) => axios.get('https://api.spacexdata.com/v3/missions').then((response) => {
    //   const books = response.data;
    dispatch(fetchRocketSuccess(response.data));
    // console.log('books from API is ', books)
    //   return books
  });
}
