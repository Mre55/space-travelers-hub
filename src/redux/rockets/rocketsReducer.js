// *Constants
const FETCH_ROCKET = 'FETCH_ROCKET';
const RESERVE_ROCKET = 'RESERVE_ROCKET';
const CANCEL_ROCKET = 'CANCEL_ROCKET';

// *Actions
// Store data
const storeRockets = (payload) => ({
  type: FETCH_ROCKET,
  payload,
});

export const reserveRocket = (payload) => ({
  type: RESERVE_ROCKET,
  payload,
});

export const cancelReservation = (payload) => ({
  type: CANCEL_ROCKET,
  payload,
});

// Fetch data from API
export const fetchingData = () => async (dispatch) => {
  const data = await fetch('https://api.spacexdata.com/v3/rockets', {
    method: 'GET',
  });
  try {
    const rockets = await data.json();
    dispatch(storeRockets(rockets));
  } catch (error) {
    console.error(error);
  }
};

// *Reducer function
const rocketsReducer = (state = [], action) => {
  switch (action.type) {
    case FETCH_ROCKET:
      return [
        action.payload.map((rocket) => ({
          flickrImages: rocket.flickr_images[0],
          rocketName: rocket.rocket_name,
          description: rocket.description,
          id: rocket.rocket_id,
          reserve: false,
        })),
      ];
    case RESERVE_ROCKET:
      return [
        state[0].map((rocket) => (rocket.id === action.payload
          ? { ...rocket, reserve: true } : rocket)),
      ];
    case CANCEL_ROCKET:
      return [
        state[0].map((rocket) => (rocket.id === action.payload
          ? { ...rocket, reserve: false } : rocket)),
      ];
    default:
      return state;
  }
};

export default rocketsReducer;
