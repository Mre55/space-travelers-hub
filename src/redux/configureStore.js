import { applyMiddleware, combineReducers, createStore } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import missionsReducer from './missions/missions';
import rocketsReducer from './rockets/rocketsReducer';

const reducer = combineReducers({ missionsReducer, rocketsReducer });

const store = createStore(reducer, applyMiddleware(thunk, logger));

export default store;
