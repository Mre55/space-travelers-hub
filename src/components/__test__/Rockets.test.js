import configureStore from 'redux-mock-store';

import { reserveRocket, storeRockets, cancelReservation } from '../../redux/rockets/rocketsReducer';

const middlewares = []
const mockStore = configureStore(middlewares)

it('should dispatch Reserve Rocket action', () => {

  const initialState = {}
  const store = mockStore(initialState)

  // Dispatch the action
  store.dispatch(reserveRocket())

  // Test if your store dispatched the expected actions
  const actions = store.getActions()
  const expectedPayload = { type: 'RESERVE_ROCKET' }
  expect(actions).toEqual([expectedPayload])
})

it('should dispatch cancel Reservation action', () => {

  const initialState = {}
  const store = mockStore(initialState)

  // Dispatch the action
  store.dispatch(cancelReservation())

  // Test if your store dispatched the expected actions
  const actions = store.getActions()
  const expectedPayload = { type: 'CANCEL_ROCKET' }
  expect(actions).toEqual([expectedPayload])
})
