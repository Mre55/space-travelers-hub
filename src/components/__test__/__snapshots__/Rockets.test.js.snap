// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo snapshot tests renders correctly 1`] = `
<div>
  <div
    className="grid m-12"
  >
    <div>
      <div
        className="flex space-x-10"
      >
        <div>
          <img
            alt="Rocket"
            className="flex-none w-80"
          />
        </div>
        <div
          className="flex-none flex flex-col space-y-4 w-8/12"
        >
          <div>
            <div
              className="text-3xl"
            />
          </div>
          <div>
            <div
              className="gap-4"
            >
              <div
                className="bg-teal-600 text-sm text-white rounded mt-1 text-center mr-4 align-middle h-5 w-20 float-left"
              >
                Reserved
              </div>
              <div>
                description
              </div>
            </div>
            ) : ( description )
          </div>
          <div>
            <button
              className="bg-inherit border-solid border-2 text-lg text-gray-500 border-gray-400 p-2 rounded"
              onClick={[Function]}
              type="button"
            >
              Cancel Reservation
            </button>
            ) : (
            <button
              className="bg-blue-600 text-white p-2 w-40 text-lg  rounded"
              onClick={[Function]}
              type="button"
            >
              Reserve Rocket
            </button>
            )
          </div>
        </div>
      </div>
      <br />
    </div>
  </div>
</div>
`;
